### PR TITLE
Add the "End of File" blank line formatting option #4641

### PR DESCRIPTION
**Pull request by junichi11**
- https://github.com/apache/netbeans/issues/4641
- Add the "End of File" blank line formatting option

nb-php-gh-4641-formatting-options

- Add the new hint for PSR-12 [Files](https://www.php-fig.org/psr/psr-12/#22-files)

nb-php-psr12-hint-options

nb-php-psr12-hint-2

nb-php-psr12-hint-1

### CODE REVIEW

File: php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java [hunk 1]
```diff
@@ -3113,6 +3118,36 @@ private String createWhitespace(DocumentOptions docOptions, int lines, int space
         return sb.toString();
     }
 
+    private void resolveNoNewlineAtEOF(BaseDocument document) {
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java:3121]: Nitpick:

```suggestion
    private void resolveNoNewLineAtEOF(BaseDocument document) {
```

**Review by tmysik — APPROVED** [review]
Looks good to me

**Comment by tmysik** [comment]
@junichi11 Just one note - aren't we able to sort PSR hints in the dialog? I mean, currently, we have PSR-12 before PSR-4. It would be great, if we could sorry it "by PSR number".

File: php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java:3121]: Will fix it.

**Comment by junichi11** [comment]
@tmysik Thank you for your review!

> Just one note - aren't we able to sort PSR hints in the dialog? I mean, currently, we have PSR-12 before PSR-4. It would be great, if we could sorry it "by PSR number".

I was also wondering about it.
They are sorted in `HintsPanel`. I have no ideas at the moment...

https://github.com/apache/netbeans/blob/68621165502b00599adf4622e0e7a14fc0565cb1/ide/csl.api/src/org/netbeans/modules/csl/hints/infrastructure/HintsPanel.java#L106-L125

File: php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/indent/TokenFormatter.java]: Fixed. Thanks!

**Comment by junichi11** [comment]
Or, add space after `-`

```
csl-hints/text/x-php5/hints/psr0=PSR- 0: Autoloading Standard
csl-hints/text/x-php5/hints/psr1=PSR- 1: Basic Coding Standard
csl-hints/text/x-php5/hints/psr4=PSR- 4: Autoloader
csl-hints/text/x-php5/hints/psr12=PSR-12: Extended Coding Style
```
nb-php-gh-4641-sort-psrs

**Comment by tmysik** [comment]
@junichi11 I see. Hmm, up to you to decide. Personally, I don't like the extra space but like the sorting 😅

**Comment by junichi11** [comment]
@tmysik 

> Personally, I don't like the extra space but like the sorting

Me, too...
I'll merge this as it is :) Thanks!

**Pull request merged**